sql/core/src/main/scala/org/apache/spark/sql/execution/QueryExecution.scala

```diff
@@ -73,12 +76,33 @@ class QueryExecution(
     sparkSession.sessionState.analyzer.executeAndCheck(logical, tracker)
   }
 
+  // SPARK-35378: Commands should be executed eagerly so that `sql("INSERT ...")` can trigger the
+  // table insertion immediately without a `.collect()`. We also need to eagerly execute non-root
+  // commands, because many commands return `GenericInternalRow` and can't be put in a query plan
+  // directly, otherwise the query engine may cast `GenericInternalRow` to `UnsafeRow` and fail.
+  lazy val commandExecuted: LogicalPlan = if (isExecutingCommand) {
+    analyzed.mapChildren(eagerlyExecuteCommands)
+  } else {
+    eagerlyExecuteCommands(analyzed)
+  }
+
+  private def eagerlyExecuteCommands(p: LogicalPlan) = p transformDown {
+    case c: Command =>
+      val qe = sparkSession.sessionState.executePlan(c, true)
+      CommandResult(
+        qe.analyzed.output,
+        qe.commandExecuted,
+        qe.executedPlan,
+        SQLExecution.withNewExecutionId(qe, name)(qe.executedPlan.executeCollect()))
+    case other => other
+  }
+
   lazy val withCachedData: LogicalPlan = sparkSession.withActive {
     assertAnalyzed()
     assertSupported()
     // clone the plan to avoid sharing the plan instance between different stages like analyzing,
     // optimizing and planning.
-    sparkSession.sharedState.cacheManager.useCachedData(analyzed.clone())
+    sparkSession.sharedState.cacheManager.useCachedData(commandExecuted.clone())
   }
 
   lazy val optimizedPlan: LogicalPlan = executePhase(QueryPlanningTracker.OPTIMIZATION) {
```

Comment:
To make sure we don't include command execution time in the optimization phase, it's better to follow `lazy val sparkPlan`

```
lazy val optimizedPlan: LogicalPlan = {
  assertCommandExecuted()
  executePhase(QueryPlanningTracker.OPTIMIZATION) {
    ...
  }
}
```

Reply:
OK